### PR TITLE
[#5180][part-1] improvement(test): Add more integration tests about access control

### PR DESCRIPTION
**Pull request by jerqi**
### What changes were proposed in this pull request?

1) Add more integration tests about access control.

2) Fix the can't create a role in authorization plugin with `CREATE_ROLE`, `MANAGE_USERS`.

3) Fix creating a role contains metalake, it won't take effect.

### Why are the changes needed?

Fix: #5180 

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Just tests.

### CODE REVIEW

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java [hunk 1]
```diff
@@ -300,21 +606,28 @@ private static void createCatalog() {
...
-            containerSuite.getHiveRangerContainer().getContainerIpAddress(),
-            HiveContainer.HDFS_DEFAULTFS_PORT,
-            schemaName.toLowerCase()));
-    String comment = "comment";
+  private static void waitForUpdatingPolicies() throws InterruptedException {
+    // After Ranger authorization, Must wait a period of time for the Ranger Spark plugin to update
+    // the policy Sleep time must be greater than the policy update interval
+    // (ranger.plugin.spark.policy.pollIntervalMs) in the
+    // `resources/ranger-spark-security.xml.template`
+    Thread.sleep(1000L);
```
Comment (jerryshao) [inline authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:614]: How can you make sure that `1000` is enough?

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java [hunk 2]
```diff
@@ -300,21 +606,28 @@ private static void createCatalog() {
...
+      field.setAccessible(true);
+      Map<String, String> writableEnv = (Map<String, String>) field.get(env);
+      if (value == null) {
+        writableEnv.remove(key);
+      } else {
+        writableEnv.put(key, value);
+      }
+    } catch (Exception e) {
+      throw new IllegalStateException("Failed to set environment variable", e);
+    }
```
Comment (jerryshao) [inline authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:631]: Can you please also add the tests about alter table, and also verify the ownership mechanism if possible?

**Review by jerryshao — COMMENTED** [review]
Just two general comments on IT, for the Ranger related code it would be better @xunliu can help to review.

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java (hunk 2 above)
Comment (jerqi) [reply authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:631]: I have added the case about `alter table`. You can see 332 Line.
The ownership is related to rename/delete the metadata object. I will add it later.

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java (hunk 1 above)
Comment (jerqi) [reply authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:614]: It's twice of update interval.

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java [hunk 3]
```diff
@@ -122,6 +143,8 @@ public void startIntegrationTest() throws Exception {
 
     generateRangerSparkSecurityXML();
 
+    setEnv(HADOOP_USER_NAME, "test");
```
Comment (xunliu) [inline authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:146]: I think `setEnv()` function will change the OS environment variable. Maybe affected other IT case. So I think we should use environment variable in Gradle script.

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java [hunk 4]
```diff
@@ -204,52 +229,327 @@ public void stop() {
...
+            metalakeName,
+            Lists.newArrayList(
+                Privileges.UseSchema.allow(),
+                Privileges.CreateSchema.allow(),
+                Privileges.CreateTable.allow(),
+                Privileges.SelectTable.allow(),
+                Privileges.ModifyTable.allow()));
+    String userName1 = System.getenv(HADOOP_USER_NAME);
+    metalake.createRole(readWriteRole, Collections.emptyMap(), Lists.newArrayList(securableObject));
+    metalake.grantRolesToUser(Lists.newArrayList(readWriteRole), userName1);
```
Comment (xunliu) [inline authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:310]: I think we need to add some `AccessControlException` tests before the grant role.

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java [hunk 5]
```diff
@@ -204,52 +229,327 @@ public void stop() {
...
+
+    // Clean up
+    catalog.asSchemas().dropSchema(schemaName, true);
+    metalake.deleteRole(roleName);
+  }
+
+  @Test
+  void testCreateTable() throws InterruptedException {
+    // First, create a role for creating a database and grant role to the user
+    String createSchemaRole = "createSchemaRole";
```
Comment (xunliu) [inline authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:258]: I think we can use `currentFunName()` replace `"createSchemaRole"`, keep role name unique.
The same can be done for role names in other test functions

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java [hunk 6]
```diff
@@ -204,52 +229,327 @@ public void stop() {
...
+        SecurableObjects.parse(
+            String.format("%s", catalogName),
+            MetadataObject.Type.CATALOG,
+            Lists.newArrayList(Privileges.UseCatalog.allow(), Privileges.CreateSchema.allow()));
+    metalake.createRole(roleName, Collections.emptyMap(), Lists.newArrayList(securableObject));
+
+    // Granted this role to the spark execution user `HADOOP_USER_NAME`
+    String userName1 = System.getenv(HADOOP_USER_NAME);
+    metalake.grantRolesToUser(Lists.newArrayList(roleName), userName1);
+
```
Comment (xunliu) [inline authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:499]: I think it better to add some Spark tests in the here, confirm the role is created successfully.

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java [hunk 7]
```diff
@@ -204,52 +229,327 @@ public void stop() {
...
+    // Delete the role
+    metalake.deleteRole(roleName);
+
```
Comment (xunliu) [inline authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:502]: I think it better to add some Spark tests in the here, confirm the role is deleted successfully.

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java [hunk 8]
```diff
@@ -204,52 +229,327 @@ public void stop() {
...
+    // Create the role again
+    metalake.createRole(roleName, Collections.emptyMap(), Lists.newArrayList(securableObject));
+
+    // Grant the role again
+    metalake.grantRolesToUser(Lists.newArrayList(roleName), userName1);
+
```
Comment (xunliu) [inline authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:508]: I think it better to add some Spark tests in the here, confirm the role is created successfully again.

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java (hunk 4 above)
Comment (jerqi) [reply authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:310]: These cases have been tested by the method `testCreateTable`.

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java (hunk 5 above)
Comment (jerqi) [reply authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:258]: OK.

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java (hunk 3 above)
Comment (jerqi) [reply authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java]: Removed.

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java (hunk 6 above)
Comment (jerqi) [reply authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:507]: I added.

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java (hunk 7 above)
Comment (jerqi) [reply authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:515]: I added.

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java (hunk 8 above)
Comment (jerqi) [reply authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:525]: I added.

File: build.gradle.kts [hunk 9]
```diff
@@ -163,7 +163,11 @@ allprojects {
       // Default use MiniGravitino to run integration tests
       param.environment("GRAVITINO_ROOT_DIR", project.rootDir.path)
       param.environment("IT_PROJECT_DIR", project.buildDir.path)
-      param.environment("HADOOP_USER_NAME", "anonymous")
+      if (project.name == "authorization-ranger") {
+        param.environment("HADOOP_USER_NAME", "test")
+      } else {
+        param.environment("HADOOP_USER_NAME", "anonymous")
+      }
```
Comment (xunliu) [inline build.gradle.kts:170]: I think we can user-define environment variable `HADOOP_USER_NAME`  in the submodule

```
      // If the environment variable `HADOOP_USER_NAME` is not customized in submodule,
      // then set it to "anonymous"
      if (param.environment["HADOOP_USER_NAME"] == null) {
        param.environment("HADOOP_USER_NAME", "anonymous")
      }
```

and add these code in the `authorization-ranger` submodule 
```
tasks.test {
  ......

  doFirst {
    environment("HADOOP_USER_NAME", "test")
  }
}
```

Comment (jerqi) [reply build.gradle.kts]: Done.

**Review by xunliu — APPROVED** [review]
LGTM

**Pull request merged**